### PR TITLE
Out-of-process inspection of details needed for Mono callstack generation

### PR DESCRIPTION
**Pull request by PJayB**
This exposes the necessary details for symbolicating managed callstacks out-of-process. For x64, this also exposes the Function Access Tables maintained by Mono for purposes of stack unwinding out-of-process. 

New APIs:
+mono_unity_oop_init // Sets up out-of-process callbacks
+mono_unity_oop_iterate_dynamic_function_access_tables64 // Iterates over the FATs (x64 only)
+mono_unity_oop_get_dynamic_function_access_table64 // Iterates over the FATs (x64 only)
+mono_unity_oop_get_stack_frame_details // Given an address, return stack frame human-readable name
+mono_unity_lock_dynamic_function_access_tables64* // Locks the FATs to prevent corruption by other threads (x64 only)
+mono_unity_unlock_dynamic_function_access_tables64* // Unlocks the FATs (x64 only)

(* These don't obey the oop_ naming convention as they can't be called from external processes. The crashing process must call lock() before allowing any external process to inspect the FATs.)

To Do/Known Issues:
* oop_init stores callbacks as globals, which isn't great, but made the code so much simpler. Feedback appreciated.
* Expose a few more details from get_stack_frame_details: return type, parameters, whatever we can.
* Testing generated names for nested classes, generics, etc.
* Comments, docs.

### CODE REVIEW

File: mono/metadata/oop.c [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+static OutOfProcessMono g_oop = { NULL, NULL };
+
+#define OFFSET_MEMBER(type, base, member) ((gpointer)((gchar*)(base) + offsetof(type, member)))
+
+void read_exception(const void* address, gsize size)
+{
+    if (g_oop.readException)
+        g_oop.readException(address, size, g_oop.userData);
+    else
+        abort();
```
Comment (joshpeterson) [inline mono/metadata/oop.c]: I think this should be `g_assert_not_reached` instead of `abort`, as the former will give us a stack trace and some diagnostic information if we reach this line of code. I don't think `abort` will do that on some platforms.

File: mono/metadata/oop.c [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+            if ((gint8*)addr >= (gint8*)ji.code_start
+                && (gint8*)addr < (gint8*)ji.code_start + ji.code_size) {
+                return chunk.data[pos];
+            }
+
+            /* If we find a non-tombstone element which is already
+            beyond what we're looking for, we have to end the
+            search. */
+            if ((gint8*)addr < (gint8*)ji.code_start)
+                goto not_found;
```
Comment (joshpeterson) [inline mono/metadata/oop.c]: If possible, I would prefer to avoid adding new `goto` statements. Can we just have `return NULL` here instead?

File: msvc/libmonoruntime.vcxproj [hunk 3]
```diff
@@ -28,6 +28,9 @@
   <ItemGroup>
     <None Include="..\mono\metadata\Makefile.am" />
   </ItemGroup>
+  <ItemGroup>
+    <ClCompile Include="..\mono\metadata\oop.c" />
+  </ItemGroup>
```
Comment (joshpeterson) [inline msvc/libmonoruntime.vcxproj]: This file should be added to the libmonoruntime-common.targets file instead of the msvc/libmonoruntime.vcxproj file. Currently, we're separating the list of files to compile from the project which indicates how to compile them. That might change in the future, but let's stick to it for now.

File: .gitmodules [hunk 4]
```diff
@@ -29,7 +29,7 @@
 	url = git://github.com/mono/NUnitLite.git
 [submodule "external/buildscripts"]
 	path = external/buildscripts
-	url = git@github.com:Unity-Technologies/monobuildtools.git
+	url = https://github.com/Unity-Technologies/monobuildtools.git
```
Comment (joshpeterson) [inline .gitmodules]: I think that you will need to rebase on unity-2017-02, as @joncham has made the external/buildscripts directory no longer a submodule.

**Comment by joshpeterson** [comment]
Generally, this looks good, thanks! I've left a few comments, and it looks like there are some minor conflicts to resolve now.

Are these changes going to be tested in the Unity repo?

File: mono/metadata/oop.c (hunk 2 above)
Comment (PJayB) [inline mono/metadata/oop.c]: Agreed. This is a clone of jit_info_table_find (adapted to be OOP) and I missed that. As a result of this, I've also added comments to both functions that one is derived from the other.

File: mono/metadata/oop.c (hunk 1 above)
Comment (PJayB) [inline mono/metadata/oop.c]: Good thinking. I'll assert that g_oop.readException is not NULL first, then replace abort() with eg_unreachable(). The functionality is the same as what you proposed but you get a more meaningful assert message.

**Comment by PJayB** [comment]
Thanks! Updated the branch with your feedback -- except the merge, which I'll get to in a moment. These changes will be tested in the Unity repo (once I figure out how to do that :))

File: mono/metadata/oop.c [hunk 5]
```diff
@@ -69,10 +69,11 @@ static OutOfProcessMono g_oop = { NULL, NULL };
 
 void read_exception(const void* address, gsize size)
 {
+    g_assert(g_oop.readException);
     if (g_oop.readException)
         g_oop.readException(address, size, g_oop.userData);
     else
-        abort();
+        eg_unreachable();
```
Comment (joshpeterson) [inline mono/metadata/oop.c]: Generally, `g_assert_not_reached` is used instead of `eg_unreachable`. I would prefer that.

**Comment by PJayB** [comment]
Rebased on top of unity-2017-02.

File: .gitmodules (hunk 4 above)
Comment (joshpeterson) [inline .gitmodules]: I wonder why this change is still, here, even after you rebased onto unity-2017-02. Let's wait for input from @joncham before we merge this.

File: mono/metadata/oop.c [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+#define OFFSET_MEMBER(type, base, member) ((gpointer)((gchar*)(base) + offsetof(type, member)))
+
+void read_exception(const void* address, gsize size)
+{
+    g_assert(g_oop.readException);
+    if (g_oop.readException)
+        g_oop.readException(address, size, g_oop.userData);
+    else
+        eg_unreachable();
```
Comment (joshpeterson) [inline mono/metadata/oop.c]: It looks like my last comment here might have been lost with the rebase. I think this should be `g_assert_not_reached`, as that is what is commonly used in Mono for this purpose.

Comment (PJayB) [inline mono/metadata/oop.c]: g_assert_not_reached calls eg_unreachable. A regular g_assert handles the actual error message a few lines above. I am tempted to change this to:

```g_assert(g_oop.readException);
g_oop.readException(...);```

...and not even bother with the branch.

Comment (joshpeterson) [inline mono/metadata/oop.c]: That actually might be cleaner.

**Comment by PJayB** [comment]
Switching to a different branch (based off staging). I'll open a new PR.
